src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,515 @@
+                                               filtered_sections_to_insert, filtered_sections_to_delete},
+                     {});
+            }
+        }
+        else {
+            m_cb(SectionedResultsChangeSet{converted_insertions, converted_modifications, converted_deletions,
+                                           section_changes.first, section_changes.second},
+                 {});
+        }
+
+        REALM_ASSERT(m_sectioned_results.m_results.is_valid());
+        m_prev_row_to_index_path = m_sectioned_results.m_row_to_index_path;
+    }
+    void error(std::exception_ptr ptr)
+    {
+        m_cb({}, ptr);
+    }
+
+    std::pair<IndexSet, IndexSet> calculate_sections_to_insert_and_delete() REQUIRES(m_sectioned_results.m_mutex)
+    {
+        IndexSet sections_to_insert, sections_to_remove;
+
+        for (auto& [key, section] : m_sectioned_results.m_sections) {
+            if (m_sectioned_results.m_previous_key_to_index_lookup.find(key) ==
+                m_sectioned_results.m_previous_key_to_index_lookup.end()) {
+                sections_to_insert.add(section.index);
+            }
+        }
+
+        for (auto& [key, index] : m_sectioned_results.m_previous_key_to_index_lookup) {
+            if (m_sectioned_results.m_sections.find(key) == m_sectioned_results.m_sections.end()) {
+                sections_to_remove.add(index);
+            }
+        }
+
+        return {sections_to_insert, sections_to_remove};
+    }
+
+private:
+    SectionedResultsNotificatonCallback m_cb;
+    SectionedResults& m_sectioned_results;
+    std::vector<std::pair<size_t, size_t>> m_prev_row_to_index_path;
+    // When set change notifications will be filtered to only deliver
+    // change indices refering to the supplied section key.
+    util::Optional<Mixed> m_section_filter;
+};
+
+template <typename StringType>
+void create_buffered_key(Mixed& key, std::unique_ptr<char[]>& buffer, StringType value)
+{
+    if (value.size() == 0) {
+        key = StringType("", 0);
+    }
+    else {
+        buffer = std::make_unique<char[]>(value.size());
+        std::strncpy(buffer.get(), value.data(), value.size());
+        key = StringType(buffer.get(), value.size());
+    }
+}
+
+ResultsSection::ResultsSection()
+    : m_parent(nullptr)
+    , m_key(Mixed())
+{
+}
+
+ResultsSection::ResultsSection(SectionedResults* parent, Mixed key)
+    : m_parent(parent)
+{
+    // Give the ResultsSection its own copy of the string data
+    // to counter the event that the m_previous_str_buffers, m_current_str_buffers
+    // no longer hold a reference to the data.
+    if (key.is_type(type_String, type_Binary)) {
+        key.is_type(type_String) ? create_buffered_key(m_key, m_key_buffer, key.get_string())
+                                 : create_buffered_key(m_key, m_key_buffer, key.get_binary());
+    }
+    else {
+        m_key = key;
+    }
+}
+
+bool ResultsSection::is_valid() const
+{
+    util::CheckedUniqueLock lock(m_parent->m_mutex);
```

Comment:
I think `ResultsSection` doesn't need any locking? Constructing a `ResultsSection` requires evaluating the `SectionedResults`, and a frozen SectionedResults should be immutable after it has been evaluated.

Reply:
The reason for the lock here is that calls to `m_parent` member variables & methods are guarded by the mutex. If we want to remove the use of the lock I think we'd end up having to copy the required `Section` from `m_parent->m_sections`. Which would probably end up causing issues because it would not let the `ResultsSection` be lazily evaluated.